clang/lib/CConv/ArrayBoundsInferenceConsumer.cpp

```diff
@@ -678,6 +729,9 @@ class ComparisionVisitor : public RecursiveASTVisitor<ComparisionVisitor> {
   std::set<BoundsKey> &PB;
   // Helper objects.
   ConstraintResolver *CR;
+  // Current statement: The statement to which the processing
+  // node belongs. This is to avoid walking the AST.
+  Stmt *CS;
```

Comment:
I don't understand why this is needed. If you are running the visitor you know what the current statement is. If you have a visitor function calling another function, you can just pass that statement as an argument. This sort of programming pattern seems brittle to me.

Reply:
The visitor methods are called by clang's `RecursiveASTVisitor`, in the `visit**` methods you get to see only the node on which you are currently operating. Since `RecursiveASTVisitor` traverses AST in preorder (https://github.com/plum-umd/checkedc-clang/blob/BigRefactor/clang/include/clang/AST/RecursiveASTVisitor.h#L147), this is a reasonable way to keep track of the parent Statement.

Another way to do this, is to given an `Expr` node,  traverse upwards the AST to find the nearest parent Statement.
However, clang's support for getting parents of an ASTNode is not that great.

Reply:
Seems brittle, since you have to remember to set this properly. But if this is the best approach we can go ahead. Merge at will!

Reply:
I have added comments in the code for clarification so that it would be easy to remember why we are having this (brittle!?) code :)